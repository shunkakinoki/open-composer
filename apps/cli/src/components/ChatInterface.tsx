import { TextAttributes } from "@opentui/core";
import { useKeyboard } from "@opentui/react";
import type React from "react";
import { useState } from "react";

interface ChatMessage {
  id: string;
  content: string;
  sender: "user" | "agent";
  agent?: string;
  timestamp: Date;
}

interface ChatInterfaceProps {
  messages: ChatMessage[];
  onSendMessage: (message: string) => void;
}

export const ChatInterface: React.FC<ChatInterfaceProps> = ({
  messages,
  onSendMessage,
}) => {
  const [input, setInput] = useState("");

  useKeyboard((key) => {
    if (key.name === "return") {
      if (input.trim()) {
        onSendMessage(input.trim());
        setInput("");
      }
    } else if (key.name === "backspace" || key.name === "delete") {
      setInput((prev) => prev.slice(0, -1));
    } else if (key.sequence && key.sequence.length === 1 && !key.ctrl && !key.meta) {
      setInput((prev) => prev + key.sequence);
    }
  });

  const formatTime = (date: Date) => {
    return date.toLocaleTimeString("en-US", {
      hour12: false,
      hour: "2-digit",
      minute: "2-digit",
    });
  };

  const getAgentIcon = (agent?: string) => {
    switch (agent) {
      case "claude-code":
        return "🤖";
      case "codex-nation":
        return "📝";
      case "cursor-agent":
        return "🖱️";
      default:
        return "🎼";
    }
  };

  return (
    <box style={{ flexDirection: "column", padding: 1 }}>
      <text
        content="💬 Chat Interface"
        style={{ fg: "green", attributes: TextAttributes.BOLD }}
      />

      <box style={{ flexDirection: "column", flexGrow: 1, marginTop: 1 }}>
        {messages.map((message) => (
          <box key={message.id} style={{ flexDirection: "column", marginBottom: 1 }}>
            <box>
              <text
                content={`${message.sender === "user" ? "👤" : getAgentIcon(message.agent)} ${message.sender === "user" ? "You" : message.agent || "Agent"}`}
                style={{ fg: message.sender === "user" ? "cyan" : "yellow" }}
              />
              <text
                content={` ${formatTime(message.timestamp)}`}
                style={{ fg: "gray" }}
              />
            </box>
            <text content={` ${message.content}`} />
          </box>
        ))}
      </box>

      <box style={{ border: true, borderColor: "blue", marginTop: 1 }}>
        <text content="💭 " style={{ fg: "blue" }} />
        <text content={input} />
        <text content="_" style={{ fg: "gray" }} />
      </box>

      <text content="Type your message and press Enter..." style={{ fg: "gray" }} />
    </box>
  );
};
