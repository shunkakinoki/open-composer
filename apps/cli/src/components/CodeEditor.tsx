// biome-ignore assist/source/organizeImports: False positive
import type React from "react";
import { Box, Text } from "ink";

interface CodeEditorProps {
  currentFile?: string;
}

export const CodeEditor: React.FC<CodeEditorProps> = ({ currentFile }) => {
  return (
    <Box flexDirection="column" padding={1}>
      <Text bold color="magenta">
        📝 Code Editor
      </Text>

      <Box flexDirection="column" marginTop={1}>
        {currentFile ? (
          <>
            <Text color="blue">📄 {currentFile}</Text>
            <Box
              borderStyle="single"
              borderColor="gray"
              padding={1}
              marginTop={1}
              flexGrow={1}
            >
              <Text color="gray" dimColor>
                Code content will appear here...
              </Text>
            </Box>
          </>
        ) : (
          <Box
            borderStyle="single"
            borderColor="gray"
            padding={2}
            flexGrow={1}
            justifyContent="center"
            alignItems="center"
          >
            <Text color="gray" dimColor>
              No file selected
            </Text>
          </Box>
        )}
      </Box>

      <Box marginTop={1}>
        <Text bold color="yellow">
          🔧 File Tree
        </Text>
      </Box>

      <Box flexDirection="column" marginTop={1}>
        <Text color="gray">📁 src/</Text>
        <Text color="gray">{"  "}📄 index.ts</Text>
        <Text color="gray">{"  "}📁 components/</Text>
        <Text color="gray">{"    "}📄 ComposerApp.tsx</Text>
        <Text color="gray">{"    "}📄 ChatInterface.tsx</Text>
      </Box>
    </Box>
  );
};
