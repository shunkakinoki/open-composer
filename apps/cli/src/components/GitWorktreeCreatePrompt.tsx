import { TextAttributes } from "@opentui/core";

import { useKeyboard } from "@opentui/react"; 
import type React from "react";
import { useState } from "react";

export interface GitWorktreeCreateOptions {
  path: string;
  ref?: string;
  branch?: string;
  force: boolean;
  detach: boolean;
  noCheckout: boolean;
  branchForce: boolean;
}

interface GitWorktreeCreatePromptProps {
  onSubmit: (options: GitWorktreeCreateOptions) => void;
  onCancel?: () => void;
  defaultPath?: string;
}

export const GitWorktreeCreatePrompt: React.FC<
  GitWorktreeCreatePromptProps
> = ({ onSubmit, onCancel, defaultPath = "" }) => {
  const [currentField, setCurrentField] = useState<
    "path" | "ref" | "branch" | "options"
  >("path");
  const [options, setOptions] = useState<GitWorktreeCreateOptions>({
    path: defaultPath,
    ref: "",
    branch: "",
    force: false,
    detach: false,
    noCheckout: false,
    branchForce: false,
  });
  

  const updateOption = <K extends keyof GitWorktreeCreateOptions>(
    key: K,
    value: GitWorktreeCreateOptions[K],
  ) => {
    setOptions((prev) => ({ ...prev, [key]: value }));
  };

  const nextField = () => {
    if (currentField === "path") setCurrentField("ref");
    else if (currentField === "ref") setCurrentField("branch");
    else if (currentField === "branch") setCurrentField("options");
  };

  const prevField = () => {
    if (currentField === "ref") setCurrentField("path");
    else if (currentField === "branch") setCurrentField("ref");
    else if (currentField === "options") setCurrentField("branch");
  };

  const handleSubmit = () => {
    if (!options.path.trim()) return; // Require at least a path

    onSubmit(options);
    process.exit(0);
  };

  const toggleOption = (
    option: keyof Pick<
      GitWorktreeCreateOptions,
      "force" | "detach" | "noCheckout" | "branchForce"
    >,
  ) => {
    updateOption(option, !options[option]);
  };

  useKeyboard(
    (key) => {
      if (key.shift && key.name === "tab") {
        prevField();
      } else if (key.name === "tab") {
        nextField();
      } else if (key.name === "return") {
        if (currentField === "options") {
          handleSubmit();
        } else {
          nextField();
        }
      } else if (key.name === "escape" || (key.ctrl && key.sequence === "c")) {
        onCancel?.();
        process.exit(0);
      } else if (key.name === "backspace" || key.name === "delete") {
        // Handle backspace for current field
        if (currentField === "path") {
          updateOption("path", options.path.slice(0, -1));
        } else if (currentField === "ref") {
          updateOption("ref", options.ref?.slice(0, -1) || "");
        } else if (currentField === "branch") {
          updateOption("branch", options.branch?.slice(0, -1) || "");
        }
      } else if (input && currentField !== "options") {
        // Handle text input for current field
        if (currentField === "path") {
          updateOption("path", options.path + input);
        } else if (currentField === "ref") {
          updateOption("ref", (options.ref || "") + input);
        } else if (currentField === "branch") {
          updateOption("branch", (options.branch || "") + input);
        }
      } else if (key.name === "up" || key.name === "down") {
        // Handle option toggling when in options field
        if (currentField === "options") {
          const optionOrder: (keyof GitWorktreeCreateOptions)[] = [
            "force",
            "detach",
            "noCheckout",
            "branchForce",
          ];
          const currentIndex = optionOrder.indexOf("force"); // Start with force
          toggleOption(
            optionOrder[currentIndex] as keyof Pick<
              GitWorktreeCreateOptions,
              "force" | "detach" | "noCheckout" | "branchForce"
            >,
          );
        }
      }
    },
    { isActive: true },
  );

  const renderField = (
    fieldName: string,
    value: string,
    placeholder: string,
    required = false,
  ) => (
    <box>
      <text content={`${fieldName}:`} style={{ fg: "cyan" }} />
      <text
        content={`${value || placeholder}${required && !value ? " *" : ""}`}
        style={{
          fg: currentField === (fieldName.toLowerCase() as "path" | "ref" | "branch" | "options") ? "green" : "gray"
        }}
      />
      {currentField === (fieldName.toLowerCase() as typeof currentField) && (
        <text content="_" style={{ fg: "gray" }} />
      )}
    </box>
  );

  const renderOption = (name: string, value: boolean, keyBinding?: string) => (
    <box>
      <text
        content={`[${value ? "●" : "○"}] ${name}${keyBinding ? ` (${keyBinding})` : ""}`}
        style={{ fg: value ? "green" : "red" }}
      />
    </box>
  );

  return (
    <box flexDirection="column" padding={2}>
      <text content="🚀 Create Git Worktree" style={{ fg: "cyan", attributes: TextAttributes.BOLD }} />

      <box marginTop={1} marginBottom={1}>
        <text content="Create a new git worktree with interactive configuration." />
      </box>

      <box marginTop={2} marginBottom={1}>
        <text content="Configuration:" style={{ attributes: TextAttributes.BOLD }} />
      </box>

      {renderField("Path", options.path, "Enter worktree path", true)}
      {renderField("Ref", options.ref || "", "Optional commit/branch ref")}
      {renderField("Branch", options.branch || "", "Optional branch name")}

      <box marginTop={2} marginBottom={1}>
        <text content="Options:" style={{ attributes: TextAttributes.BOLD }} />
      </box>

      <box flexDirection="column">
        {renderOption("Force", options.force, "F")}
        {renderOption("Detach", options.detach, "D")}
        {renderOption("No Checkout", options.noCheckout, "N")}
        {renderOption("Branch Force", options.branchForce, "B")}
      </box>

      <box marginTop={2}>
        <text
          content="Tab/Shift+Tab: Navigate • Enter: Next/Submit • ↑/↓: Toggle options • Esc: Cancel"
          style={{ fg: "gray" }}
        />
      </box>

      <box marginTop={1}>
        <text
          content={!options.path.trim() ? "⚠️  Path is required" : "✅ Ready to create worktree"}
          style={{ fg: "yellow" }}
        />
      </box>
    </box>
  );
};
