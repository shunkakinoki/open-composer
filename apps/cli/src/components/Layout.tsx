// biome-ignore assist/source/organizeImports: False positive
import type React from "react";
import { Box, Text } from "ink";
import { CLI_VERSION } from "../lib/version.js";

interface LayoutProps {
  children: React.ReactNode;
}

export const Layout: React.FC<LayoutProps> = ({ children }) => {
  return (
    <Box flexDirection="column" height="100%">
      <Box borderStyle="single" borderColor="blue" justifyContent="center">
        <Text color="blue" bold>
          ðŸŽ¼ Open Composer CLI v{CLI_VERSION}
        </Text>
      </Box>

      <Box flexGrow={1}>{children}</Box>

      <Box
        borderStyle="single"
        borderColor="gray"
        justifyContent="space-between"
        flexDirection="row"
      >
        <Text color="gray">Ready</Text>
        <Text color="gray">Ctrl+C to exit</Text>
      </Box>
    </Box>
  );
};
