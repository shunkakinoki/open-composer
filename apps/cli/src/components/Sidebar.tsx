import { TextAttributes } from "@opentui/core";
import type React from "react";

interface Worktree {
  name: string;
  path: string;
  branch: string;
  active: boolean;
}

interface SidebarProps {
  worktrees: Worktree[];
  currentBranch: string;
}

export const Sidebar: React.FC<SidebarProps> = ({
  worktrees,
  currentBranch,
}) => {
  return (
    <box style={{ flexDirection: "column", padding: 1 }}>
      <text
        content="📂 Workspaces"
        style={{ fg: "blue", attributes: TextAttributes.BOLD }}
      />

      <box style={{ flexDirection: "column", marginTop: 1 }}>
        {worktrees.map((worktree) => (
          <box key={worktree.name} style={{ marginY: 0 }}>
            <text
              content={`${worktree.active ? "●" : "○"} ${worktree.name}`}
              style={{ fg: worktree.active ? "green" : "gray" }}
            />
          </box>
        ))}
      </box>

      <box style={{ marginTop: 2 }}>
        <text
          content="🌿 Branches"
          style={{ fg: "yellow", attributes: TextAttributes.BOLD }}
        />
      </box>

      <box style={{ flexDirection: "column", marginTop: 1 }}>
        <text content={`● ${currentBranch}`} style={{ fg: "green" }} />
        <text content="+ New worktree" style={{ fg: "gray" }} />
      </box>

      <box style={{ marginTop: 2 }}>
        <text
          content="🤖 Agents"
          style={{ fg: "magenta", attributes: TextAttributes.BOLD }}
        />
      </box>

      <box style={{ flexDirection: "column", marginTop: 1 }}>
        <text content="● claude-code" style={{ fg: "green" }} />
        <text content="○ codex-nation" style={{ fg: "gray" }} />
        <text content="○ cursor-agent" style={{ fg: "gray" }} />
      </box>
    </box>
  );
};
