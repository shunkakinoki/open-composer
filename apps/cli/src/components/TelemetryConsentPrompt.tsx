import { Box, Text, useApp, useInput } from "ink";
import type React from "react";
import { useState } from "react";

interface TelemetryConsentPromptProps {
  onConsent: (consent: boolean) => void;
  onCancel?: () => void;
}

export const TelemetryConsentPrompt: React.FC<TelemetryConsentPromptProps> = ({
  onConsent,
  onCancel,
}) => {
  const [selected, setSelected] = useState<"yes" | "no">("yes");
  const { exit } = useApp();

  useInput(
    (input, key) => {
      if (key.leftArrow || key.rightArrow) {
        setSelected(selected === "yes" ? "no" : "yes");
      } else if (key.return) {
        onConsent(selected === "yes");
        exit(); // Exit the Ink app after handling consent
      } else if (key.escape || (key.ctrl && input === "c")) {
        onCancel?.();
        exit(); // Exit the Ink app after handling cancellation
      }
    },
    { isActive: true },
  );

  return (
    <Box flexDirection="column" padding={2}>
      <Text bold color="cyan">
        üîí Open Composer - Privacy Notice
      </Text>

      <Box marginTop={1} marginBottom={1}>
        <Text>
          Open Composer respects your privacy and is committed to protecting
          your data.
        </Text>
      </Box>

      <Box marginTop={1} marginBottom={1}>
        <Text bold>üìä Telemetry Collection (Optional)</Text>
      </Box>

      <Box marginTop={1} marginBottom={1}>
        <Text>
          We can collect anonymous usage statistics to help improve Open
          Composer.
        </Text>
      </Box>

      <Box marginTop={1} marginBottom={1}>
        <Text>
          This includes command usage, error reports, and performance metrics.
        </Text>
      </Box>

      <Box marginTop={1} marginBottom={1}>
        <Text>All data is anonymized and cannot be used to identify you.</Text>
      </Box>

      <Box marginTop={1} marginBottom={1}>
        <Text>
          You can change this setting anytime with: open-composer telemetry
          disable
        </Text>
      </Box>

      <Box marginTop={2}>
        <Text>Enable telemetry collection?</Text>
      </Box>

      <Box marginTop={1}>
        <Text color="gray">
          Use arrow keys to select, Enter to confirm, Esc to skip
        </Text>
      </Box>

      <Box marginTop={2}>
        <Box marginRight={2} key="yes-option">
          <Text color={selected === "yes" ? "green" : "gray"}>
            {selected === "yes" ? "‚óè" : "‚óã"} Yes, enable telemetry
          </Text>
        </Box>
        <Box key="no-option">
          <Text color={selected === "no" ? "red" : "gray"}>
            {selected === "no" ? "‚óè" : "‚óã"} No, keep disabled
          </Text>
        </Box>
      </Box>
    </Box>
  );
};
