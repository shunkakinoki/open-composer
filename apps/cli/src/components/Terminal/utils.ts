/**
 * Terminal utility functions
 * Inspired by termact: https://github.com/MasterGordon/termact/blob/main/src/utils.ts
 */

import type { TerminalStyle } from "./types.js";
import {
  bg,
  blink,
  bold,
  doubleUnderline,
  fg,
  hidden,
  inverse,
  italic,
  strikethrough,
  underline,
} from "./ansi.js";

export const applyStyle = (style: TerminalStyle): string => {
  let result = "";
  if (style.fg) result += fg(style.fg);
  if (style.bg) result += bg(style.bg);
  if (style.bold) result += bold();
  if (style.italic) result += italic();
  if (style.underline) result += underline();
  if (style.blink) result += blink();
  if (style.reverse) result += inverse();
  if (style.hidden) result += hidden();
  if (style.strikethrough) result += strikethrough();
  if (style.doubleUnderline) result += doubleUnderline();
  return result;
};

export const wrapText = (text: string, width: number): string[] => {
  const lines: string[] = [];
  const words = text.split(" ");
  let currentLine = "";

  for (const word of words) {
    if (currentLine.length + word.length + 1 <= width) {
      currentLine += (currentLine ? " " : "") + word;
    } else {
      if (currentLine) lines.push(currentLine);
      currentLine = word;
    }
  }

  if (currentLine) lines.push(currentLine);
  return lines;
};

export const truncateText = (text: string, maxLength: number): string => {
  if (text.length <= maxLength) {
    return text;
  }
  return text.slice(0, maxLength - 3) + "...";
};

export const padText = (
  text: string,
  width: number,
  align: "left" | "center" | "right" = "left",
): string => {
  if (text.length >= width) {
    return text.slice(0, width);
  }

  const padding = width - text.length;

  switch (align) {
    case "center": {
      const leftPad = Math.floor(padding / 2);
      const rightPad = padding - leftPad;
      return " ".repeat(leftPad) + text + " ".repeat(rightPad);
    }
    case "right":
      return " ".repeat(padding) + text;
    default:
      return text + " ".repeat(padding);
  }
};

export const getNtscGrayscale = (r: number, g: number, b: number): number => {
  return r * 0.299 + g * 0.587 + b * 0.114;
};
