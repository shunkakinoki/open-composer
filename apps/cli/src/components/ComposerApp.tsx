import { Box } from "ink";
import type React from "react";
import { useState } from "react";
import { ChatInterface } from "./ChatInterface.js";
import { CodeEditor } from "./CodeEditor.js";
import { Layout } from "./Layout.js";
import { Sidebar } from "./Sidebar.js";

// Utility function for creating timestamps - can be mocked in tests
export const createTimestamp = () => new Date();
export const getCurrentTime = () => Date.now();

export interface ComposerAppState {
  currentBranch: string;
  currentFile?: string;
  chatMessages: Array<{
    id: string;
    content: string;
    sender: "user" | "agent";
    agent?: string;
    timestamp: Date;
  }>;
  worktrees: Array<{
    name: string;
    path: string;
    branch: string;
    active: boolean;
  }>;
}

export const ComposerApp: React.FC = () => {
  const [state, setState] = useState<ComposerAppState>({
    currentBranch: "main",
    chatMessages: [
      {
        id: "1",
        content: "Welcome to Open Composer CLI! What would you like to build?",
        sender: "agent",
        agent: "composer",
        timestamp: createTimestamp(),
      },
    ],
    worktrees: [
      {
        name: "main",
        path: process.cwd(),
        branch: "main",
        active: true,
      },
    ],
  });

  const handleSendMessage = (message: string) => {
    const newMessage = {
      id: getCurrentTime().toString(),
      content: message,
      sender: "user" as const,
      timestamp: createTimestamp(),
    };

    setState((prev) => ({
      ...prev,
      chatMessages: [...prev.chatMessages, newMessage],
    }));

    // Mock agent response
    setTimeout(() => {
      const agentResponse = {
        id: (getCurrentTime() + 1).toString(),
        content: `I'll help you with: "${message}". Let me analyze your request...`,
        sender: "agent" as const,
        agent: "claude-code",
        timestamp: createTimestamp(),
      };

      setState((prev) => ({
        ...prev,
        chatMessages: [...prev.chatMessages, agentResponse],
      }));
    }, 1000);
  };

  return (
    <Layout>
      <Box flexDirection="row" height="100%">
        <Box width="30%" borderStyle="single" borderColor="gray">
          <Sidebar
            worktrees={state.worktrees}
            currentBranch={state.currentBranch}
          />
        </Box>

        <Box width="40%" borderStyle="single" borderColor="gray">
          <ChatInterface
            messages={state.chatMessages}
            onSendMessage={handleSendMessage}
          />
        </Box>

        <Box width="30%" borderStyle="single" borderColor="gray">
          <CodeEditor currentFile={state.currentFile} />
        </Box>
      </Box>
    </Layout>
  );
};
