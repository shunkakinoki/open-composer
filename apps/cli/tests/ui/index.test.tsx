import { describe, expect, mock, test } from "bun:test";
import { Text } from "ink";

// Mock ComposerApp utility functions before importing
const mockCreateTimestamp = mock(() => new Date("2024-01-01T10:00:00Z"));
const mockGetCurrentTime = mock(() =>
  new Date("2024-01-01T10:00:00Z").getTime(),
);

mock.module("../../src/components/ComposerApp.js", () => ({
  createTimestamp: mockCreateTimestamp,
  getCurrentTime: mockGetCurrentTime,
}));

// Mock CLI version to keep snapshots stable across version changes
mock.module("../../src/lib/version.js", () => ({
  CLI_VERSION: "0.0.0",
}));

import { GitLive } from "@open-composer/git-worktrees";
import * as Effect from "effect/Effect";
import { ChatInterface } from "../../src/components/ChatInterface.js";
import { CodeEditor } from "../../src/components/CodeEditor.js";
import { ComposerApp } from "../../src/components/ComposerApp.js";
import { GitWorktreeCreatePrompt } from "../../src/components/GitWorktreeCreatePrompt.js";
import { Layout } from "../../src/components/Layout.js";
import { Sidebar } from "../../src/components/Sidebar.js";
import { TelemetryConsentPrompt } from "../../src/components/TelemetryConsentPrompt.js";
import { GitWorktreeCli } from "../../src/lib/index.js";
import { render } from "../utils.js";

describe("Open Composer CLI", () => {
  describe("GitWorktreeCli", () => {
    test("should initialize with current directory", async () => {
      const cli = await Effect.runPromise(GitWorktreeCli.make());
      expect(cli).toBeInstanceOf(GitWorktreeCli);
    });

    test("should list worktrees", async () => {
      const cli = await Effect.runPromise(GitWorktreeCli.make());
      await expect(
        Effect.runPromise(cli.list().pipe(Effect.provide(GitLive))),
      ).resolves.toBeUndefined();
    });
  });

  describe("Component Snapshots", () => {
    test("ComposerApp renders correctly", () => {
      const { lastFrame } = render(<ComposerApp />);
      expect(lastFrame()).toMatchSnapshot();
    });

    test("ChatInterface renders correctly with messages", () => {
      const messages = [
        {
          id: "1",
          content: "Hello, how can I help you?",
          sender: "agent" as const,
          agent: "claude-code",
          timestamp: new Date("2024-01-01T10:00:00"),
        },
        {
          id: "2",
          content: "I need help with my code",
          sender: "user" as const,
          timestamp: new Date("2024-01-01T10:01:00"),
        },
      ];

      const { lastFrame } = render(
        <ChatInterface messages={messages} onSendMessage={() => {}} />,
      );
      expect(lastFrame()).toMatchSnapshot();
    });

    test("Sidebar renders correctly with worktrees", () => {
      const worktrees = [
        {
          name: "main",
          path: "/path/to/main",
          branch: "main",
          active: true,
        },
        {
          name: "feature-1",
          path: "/path/to/feature-1",
          branch: "feature-1",
          active: false,
        },
      ];

      const { lastFrame } = render(
        <Sidebar worktrees={worktrees} currentBranch="main" />,
      );
      expect(lastFrame()).toMatchSnapshot();
    });

    test("CodeEditor renders correctly with file", () => {
      const { lastFrame } = render(<CodeEditor currentFile="src/index.ts" />);
      expect(lastFrame()).toMatchSnapshot();
    });

    test("CodeEditor renders correctly without file", () => {
      const { lastFrame } = render(<CodeEditor />);
      expect(lastFrame()).toMatchSnapshot();
    });

    test("Layout renders correctly", () => {
      const { lastFrame } = render(
        <Layout>
          <Text>Test content</Text>
        </Layout>,
      );
      expect(lastFrame()).toMatchSnapshot();
    });

    test("TelemetryConsentPrompt renders correctly", () => {
      const mockOnConsent = mock(() => {});
      const mockOnCancel = mock(() => {});

      const { lastFrame } = render(
        <TelemetryConsentPrompt
          onConsent={mockOnConsent}
          onCancel={mockOnCancel}
        />,
      );
      expect(lastFrame()).toMatchSnapshot();
    });

    test("GitWorktreeCreatePrompt renders correctly", () => {
      const mockOnSubmit = mock(() => {});
      const mockOnCancel = mock(() => {});

      const { lastFrame } = render(
        <GitWorktreeCreatePrompt
          onSubmit={mockOnSubmit}
          onCancel={mockOnCancel}
        />,
      );
      expect(lastFrame()).toMatchSnapshot();
    });
  });
});
