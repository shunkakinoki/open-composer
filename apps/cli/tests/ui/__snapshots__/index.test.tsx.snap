// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Open Composer CLI Component Snapshots ComposerApp renders correctly 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                   🎼 Open Composer CLI v0.0.0                                    │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘
┌────────────────────────────┐┌──────────────────────────────────────┐┌────────────────────────────┐
│                            ││                                      ││                            │
│ 📂 Workspaces              ││ 💬 Chat Interface                    ││ 📝 Code Editor             │
│                            ││                                      ││                            │
│ ● main                     ││ 🎼 composer 10:00                    ││ ┌──────────────────────┐   │
│                            ││  Welcome to Open Composer CLI! What  ││ │                      │   │
│                            ││ would you like to build?             ││ │                      │   │
│ 🌿 Branches                ││                                      ││ │   No file selected   │   │
│                            ││                                      ││ │                      │   │
│ ● main                     ││                                      ││ │                      │   │
│ + New worktree             ││                                      ││ └──────────────────────┘   │
│                            ││                                      ││                            │
│                            ││                                      ││ 🔧 File Tree               │
│ 🤖 Agents                  ││                                      ││                            │
│                            ││                                      ││ 📁 src/                    │
│ ● claude-code              ││ ┌──────────────────────────────────┐ ││   📄 index.ts              │
│ ○ codex-nation             ││ │💭 _                              │ ││   📁 components/           │
│ ○ cursor-agent             ││ └──────────────────────────────────┘ ││     📄 ComposerApp.tsx     │
│                            ││ Type your message and press Enter... ││     📄 ChatInterface.tsx   │
│                            ││                                      ││                            │
└────────────────────────────┘└──────────────────────────────────────┘└────────────────────────────┘
┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│Ready                                                                               Ctrl+C to exit│
└──────────────────────────────────────────────────────────────────────────────────────────────────┘"
`;

exports[`Open Composer CLI Component Snapshots ChatInterface renders correctly with messages 1`] = `
"
 💬 Chat Interface

 🤖 claude-code 10:00
  Hello, how can I help you?

 👤 You 10:01
  I need help with my code


 ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
 │💭 _                                                                                            │
 └────────────────────────────────────────────────────────────────────────────────────────────────┘
 Type your message and press Enter...
"
`;

exports[`Open Composer CLI Component Snapshots Sidebar renders correctly with worktrees 1`] = `
"
 📂 Workspaces

 ● main
 ○ feature-1


 🌿 Branches

 ● main
 + New worktree


 🤖 Agents

 ● claude-code
 ○ codex-nation
 ○ cursor-agent
"
`;

exports[`Open Composer CLI Component Snapshots CodeEditor renders correctly with file 1`] = `
"
 📝 Code Editor

 📄 src/index.ts

 ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
 │                                                                                                │
 │ Code content will appear here...                                                               │
 │                                                                                                │
 └────────────────────────────────────────────────────────────────────────────────────────────────┘

 🔧 File Tree

 📁 src/
   📄 index.ts
   📁 components/
     📄 ComposerApp.tsx
     📄 ChatInterface.tsx
"
`;

exports[`Open Composer CLI Component Snapshots CodeEditor renders correctly without file 1`] = `
"
 📝 Code Editor

 ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
 │                                                                                                │
 │                                                                                                │
 │                                        No file selected                                        │
 │                                                                                                │
 │                                                                                                │
 └────────────────────────────────────────────────────────────────────────────────────────────────┘

 🔧 File Tree

 📁 src/
   📄 index.ts
   📁 components/
     📄 ComposerApp.tsx
     📄 ChatInterface.tsx
"
`;

exports[`Open Composer CLI Component Snapshots Layout renders correctly 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                   🎼 Open Composer CLI v0.0.0                                    │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘
Test content
┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│Ready                                                                               Ctrl+C to exit│
└──────────────────────────────────────────────────────────────────────────────────────────────────┘"
`;

exports[`Open Composer CLI Component Snapshots TelemetryConsentPrompt renders correctly 1`] = `
"

  🔒 Open Composer - Privacy Notice

  Open Composer respects your privacy and is committed to protecting your data.


  📊 Telemetry Collection (Optional)


  We can collect anonymous usage statistics to help improve Open Composer.


  This includes command usage, error reports, and performance metrics.


  All data is anonymized and cannot be used to identify you.


  You can change this setting anytime with: open-composer telemetry disable



  Enable telemetry collection?

  Use arrow keys to select, Enter to confirm, Esc to skip


  ● Yes, enable telemetry  ○ No, keep disabled

"
`;

exports[`Open Composer CLI Component Snapshots GitWorktreeCreatePrompt renders correctly 1`] = `
"

  🚀 Create Git Worktree

  Create a new git worktree with interactive configuration.



  Configuration:

  Path:Enter worktree path *_
  Ref:Optional commit/branch ref
  Branch:Optional branch name


  Options:

  [○] Force (F)
  [○] Detach (D)
  [○] No Checkout (N)
  [○] Branch Force (B)


  Tab/Shift+Tab: Navigate • Enter: Next/Submit • ↑/↓: Toggle options • Esc: Cancel

  ⚠️  Path is required

"
`;
